content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-2.md

```diff
@@ -0,0 +1,231 @@
+![Apply transformations](/img/apply-transformation.png)
+
+**Here is example output from the df.show :**
+
+![Example output from the df.show](/img/example-df-show.png)
+
+We transform the Dataset with the model pipeline, which will transform the features according to the pipeline, estimate and then return the predictions in a column of a new Dateset. We also create a temporary view of the new Dataset in order to execute SQL queries.
+
+![Temporary view](/img/temporary-view.png)
+
+## **4) Write messages from the transformed DStream to a Topic**
+
+The Dataset result of the query is converted to JSON RDD Strings, then the RDD sendToKafka method is used to send the JSON key-value messages to a topic (the key is null in this case).
+
+![Write messages from the transformed DStream](/img/write-messages-from-dstream.png)
+
+![Write messages from the transformed DStream 2](/img/write-messages-from-dstream-2.png)
+
+Example message values (the output for `temp.take(2)` ) are shown below:
+
+```json
+{"_id":"DL_2017-01-01_MIA_LGA_1489","dofW":7,"carrier":"DL","origin":"MIA","dest":"LGA","crsdephour":13,"crsdeptime":1315.0,"crsarrtime":1618.0,"crselapsedtime":183.0,"label":0.0,"pred_dtree":0.0}
+{"_id":"DL_2017-01-01_LGA_MIA_1622","dofW":7,"carrier":"DL","origin":"LGA","dest":"MIA","crsdephour":8,"crsdeptime":800.0,"crsarrtime":1115.0,"crselapsedtime":195.0,"label":0.0,"pred_dtree":0.0}
+```
+
+## **5) Start receiving data and processing it. Wait for the processing to be stopped.**
+
+To start receiving data, we must explicitly call start() on the StreamingContext, then call `awaitTermination` to wait for the streaming computation to finish. We use `ssc.remember` to cache data for queries.
+
+![Receive and process data](/img/receive-process-data.png)
+
+## Streaming Data Exploration
+
+Now we can query the cached streaming data in the input temporary view flights, and the predictions temporary view `flightsp`. Below we display a few rows from the flights view:
+
+![Flights view](/img/flights-view.png)
+
+Below we display the count of predicted delayed/not delayed departures by Origin:
+
+![Predicted Departures](/img/predicted-departures.png)
+
+Below we display the count of predicted delayed/not delayed departures by Destination:
+
+![Predicted Delayed](/img/predicted-delayed.png)
+
+Below we display the count of predicted delayed/not delayed departures by Origin,Destination:
+
+![Count predicted delayed](/img/count-predicted-delayed.png)
+
+## Summary
+
+In this blog post, you learned how to use a Spark machine learning model in a Spark Streaming application, and how to integrate Spark Streaming with MapR Event Streams to consume and produce messages using the Kafka API.
+
+## Code
+
+* You can download the code and data to run these examples from here: <https://github.com/caroljmcdonald/spark-ml-flightdelay>
+* [Zeppelin Notebook for the code](https://github.com/caroljmcdonald/spark-ml-flightdelay/blob/master/notebooks/sparkmlpipelineflightdelays.json)
```

Comment:
Suggest to hotlink the Github repo URL.
You can download the code and data to run these examples from [here](https://github.com/caroljmcdonald/spark-ml-flightdelay).

Reply:
done